contracts/token/ERC865/ERC865Token.sol

```diff
@@ -0,0 +1,366 @@
+    }
+
+     /**
+      * @notice Hash (keccak256) of the payload used by decreaseApprovalPreSigned
+      * @param _token address The address of the token
+      * @param _spender address The address which will spend the funds.
+      * @param _subtractedValue uint256 The amount of tokens to decrease the allowance by.
+      * @param _fee uint256 The amount of tokens paid to msg.sender, by the owner.
+      * @param _nonce uint256 Presigned transaction number.
+      */
+    function decreaseApprovalPreSignedHashing(
+        address _token,
+        address _spender,
+        uint256 _subtractedValue,
+        uint256 _fee,
+        uint256 _nonce
+    )
+        public
+        pure
+        returns (bytes32)
+    {
+        /* "59388d78": decreaseApprovalPreSignedHashing(address,address,uint256,uint256,uint256) */
+        return keccak256(bytes4(0x59388d78), _token, _spender, _subtractedValue, _fee, _nonce);
+    }
+
+    /**
+     * @notice Hash (keccak256) of the payload used by transferFromPreSigned
+     * @param _token address The address of the token
+     * @param _from address The address which you want to send tokens from.
+     * @param _to address The address which you want to transfer to.
+     * @param _value uint256 The amount of tokens to be transferred.
+     * @param _fee uint256 The amount of tokens paid to msg.sender, by the spender.
+     * @param _nonce uint256 Presigned transaction number.
+     */
+    function transferFromPreSignedHashing(
+        address _token,
+        address _from,
+        address _to,
+        uint256 _value,
+        uint256 _fee,
+        uint256 _nonce
+    )
+        public
+        pure
+        returns (bytes32)
+    {
+        /* "b7656dc5": transferFromPreSignedHashing(address,address,address,uint256,uint256,uint256) */
+        return keccak256(bytes4(0xb7656dc5), _token, _from, _to, _value, _fee, _nonce);
+    }
+
+    /**
+     * @notice Recover signer address from a message by using his signature
+     * @param hash bytes32 message, the hash is the signed message. What is recovered is the signer address.
+     * @param sig bytes signature, the signature is generated using web3.eth.sign()
+     */
+    function recover(bytes32 hash, bytes sig) public pure returns (address) {
+      bytes32 r;
+      bytes32 s;
+      uint8 v;
+
+      //Check the signature length
+      if (sig.length != 65) {
+        return (address(0));
+      }
+
+      // Divide the signature in r, s and v variables
+      assembly {
+        r := mload(add(sig, 32))
+        s := mload(add(sig, 64))
+        v := byte(0, mload(add(sig, 96)))
+      }
+
+      // Version of signature should be 27 or 28, but 0 and 1 are also possible versions
+      if (v < 27) {
+        v += 27;
+      }
+
+      // If the version is correct return the signer address
+      if (v != 27 && v != 28) {
+        return (address(0));
+      } else {
+        return ecrecover(hash, v, r, s);
```

Comment:
Is it intentional, that the sig[] contains the signature in different order than v, r, s, which is a scientific custom, and for example used by ecrecover()?